### ISSUE TITLE
take operator on memmap is >100x slower than fancy indexing

### ISSUE THREAD

**Issue by meyer89** [numpy/numpy#13827]
Working on improvements for scikit-learn, I stumbled upon a odd behavior that the take operator on a numpy.memmap can take >100x longer than normal fancy indexing.

### Reproducing code example:

Just use "pip install scikit-learn" for the example dataset with shape (70000, 784).

```python
from numpy import arange
from sklearn.datasets import fetch_openml
X_train=fetch_openml('mnist_784')['data']
indices=arange(200, dtype=int)

slice1 = X_train.take(indices, axis=0)  # slow 
slice2 = X_train[indices]  # fast
```

I do not see any difference between slice1 and slice2, even the flags are identical. However, the second operation is significantly faster than the first one, as you can see using timeit.

```
python -m timeit -n 100 -s "from numpy import arange; from sklearn.datasets import fetch_openml; X_train=fetch_openml('mnist_784')['data']; indices=arange(200, dtype=int)" "X_train[indices]"
100 loops, best of 3: 425 usec per loop
```

```
python -m timeit -n 100 -s "from numpy import arange; from sklearn.datasets import fetch_openml; X_train=fetch_openml('mnist_784')['data']; indices=arange(200, dtype=int)" "X_train.take(indices, axis=0)"
100 loops, best of 3: 490 msec per loop
```

Please note that it is **msec** with take() agains **usec** using fancy indexing.

### Error message:

No error message, but the operation is unexpected slow. Both operators should be pretty close when it comes to runtime.

### Numpy/Python version information:

1.16.4 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0]

I also tested this with an older numpy version with identical results.

1.13.3 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0]

**Comment by seberg** [comment]
In most cases (I can give the details, but...) indexing outperforms take nowadays, due to optimizing the memory order of the operation. Note that the output memory order is different in this case probably. I am surprised the speed difference is this large, but since this is a hard-disk backed file, I suppose cache line misses (i.e. wrong non-optimal order) kicks in heavily.

You would probably see that `take` with `axis=-1` has a much more similar performance. Frankly, I am not sure we are likely to optimize this any time soon, but yes, currently `take` is intrinsically slow if you index the slow axis in memory (if you are not familiar with numpy's memory layout, please ask).

**Comment by seberg** [comment]
For the sklearn case you have here, it is a bit tricky/annoying. Because if you had an array shaped `70000x2` take is probably noticeably a bit faster (no matter the memory order).

**Comment by meyer89** [comment]
I can confirm that the order of the Input matrix matters a lot.

* F_CONTIGUOUS (default for this dataset and problematic)
  * take: 490 msec per loop
  * take with axis=-1: 601 msec per loop
  * fancy: 425 usec per loop
* C_CONTIGUOUS (no problem here):
  * take: 250 usec per loop
  * fancy: 202 usec per loop

The output is C_CONTIGUOUS in *both* cases, so no difference here.

Based on the documentation, this difference can not be expected. It seems like take is just another way to use the slicing operator.

> When axis is not None, this function does the same thing as “fancy” indexing (indexing arrays using arrays); however, it can be easier to use if you need elements along a given axis. A call such as np.take(arr, indices, axis=3) is equivalent to arr[:,:,:,indices,...].

What do you think about implementing take as simple wrapper for fancy indexing, so that the same performance is obtained?

**Comment by seberg** [comment]
Well, as I posted on the other case. There are still some paths were take is quite a bit faster. Plus take has "clip", etc. modes. So yes, it could be a thing, but I am not sure it is quite as straight forward as it may appear.

**Comment by seberg** [comment]
I am going to close this, we have an open issue about it (that is super old gh-2657). Take is implemented as `fasttake` on the dtypes, and that is assumed to be C order.
